### ISSUE TITLE
Is it working with Firefox?

### ISSUE THREAD

**Issue by gjherbiet** [m13253/dns-over-https#15]
Hello, I have installed and configured doh-server. Below is a redacted config file.

```
# HTTP listen port
listen = [
    "127.0.0.1:443",
    "[::1]:443",
    "IPV4:443",
    "[IPV6]:443",
]

# TLS certification file
cert = "/usr/local/doh-server/etc/dns_over_https.pem"

# TLS key file
key = "/usr/local/doh-server/etc/dns_over_https.key"

# HTTP path for resolve application
path = "/dns-query"

# Upstream DNS resolver
# If multiple servers are specified, a random one will be chosen each time.
upstream = [
    "127.0.0.1:53",
]

# Upstream timeout
timeout = 10

# Number of tries if upstream DNS fails
tries = 3

# Only use TCP for DNS query
tcp_only = false

# Enable logging
verbose = true
```

I use Unbound on the same server as upstream.

It seems to be working when I try with curl/doh (https://github.com/curl/doh) but not when enabling DNS-over-HTTPS in Firefox (61.0): 

```
$ ./doh www.github.com https://SERVER/dns-query
www.github.com from https://SERVER/dns-query
TTL: 60 seconds
A: 192.30.253.112
A: 192.30.253.113
CNAME: github.com
CNAME: github.com
```

```
$ tail -f /var/log/doh-server/doh-server.log
[MYIP]:59738 - - [29/Jun/2018:14:06:10 +0200] "www.github.com. IN AAAA"
MYIP - - [29/Jun/2018:14:06:10 +0200] "POST /dns-query HTTP/2.0" 200 166 "" "curl-doh/1.0"
```

This is how I setup Firefox in `about:config`:

```
network.trr.bootstrapAddress;IPV4
network.trr.mode;3
network.trr.uri;https://SERVER/dns-query
```

I see some probes arriving but resolution doesn't seem to work in Firefox.

```
$ tail -f /var/log/doh-server/doh-server.log
[MYIP]:60282 - - [29/Jun/2018:13:34:04 +0200] "example.com. IN NS"
2001:a18:1:8::163 - - [29/Jun/2018:13:34:04 +0200] "POST /dns-query HTTP/2.0" 200 126 "" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.13; rv:61.0) Gecko/20100101 Firefox/61.0"
[MYIP]:60282 - - [29/Jun/2018:13:35:00 +0200] "example.com. IN NS"
2001:a18:1:8::163 - - [29/Jun/2018:13:35:00 +0200] "POST /dns-query HTTP/2.0" 200 126 "" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.13; rv:61.0) Gecko/20100101 Firefox/61.0"
```

If I set `network.trr.mode` to `2`, then it falls back to not using DoH (as it seems not working).

Do you have any idea about this? What further tests can I make?

Regards,

_Edit_: I also tried with Firefox 62.0b3 and ended up with the same results.

**Comment by m13253** [comment]
Thank you for your detailed report!

Now I am aware of this problem. But  sorry I am currently on a trip so I may be able to fix this problem later.

**Comment by m13253** [comment]
In the meanwhile before I can fix this problem, would you please turn `network.trr.useGET` to `true` (for detailed logging, since POST is not logged), and `network.trr.confirmationNS` to `skip`?

(I got the information from [this gist](https://gist.github.com/bagder/5e29101079e9ac78920ba2fc718aceec), I am not sure if it is up to date and correct.)

**Comment by m13253** [comment]
The problem is fixed in commit e19250dc9942ccea673ad4caadd6874d1e2aca18, 7e7cd1ee908de574bc358f38677fd8866720e971, and 033865e50838dfe58ea44601dc476eb977e7cc4f.

Actually this is Firefox's problem, which uses an old protocol version.
I fixed the problem by detecting Firefox's User-Agent and switch to the old protocol for it.
If you are willing to, you may also submit an issue to Firefox upstream, persuading them to switch to `application/dns-message`.
(P.S. Someone once argued with me saying “[we shouldn't have User-Agent with DoH](https://www.ietf.org/mail-archive/web/doh/current/msg00777.html)”. Facts proved that most client & server implementations are really fragile and buggy.)

Firefox periodically detects DoH server and captive portals, and cache the results for 1 minute (or 20 minutes for Firefox 61). So once you fail one of the two the detections, you lose Internet connection for 1 minute.
So I recommend you to set `network.trr.confirmationNS = skip`, since you know you typed in a valid DoH server.
Additionally, I recommend you to set `network.trr.wait-for-portal = false`, because the operating system (Windows Vista+, Linux Gnome 3.14+, Mac OS X 10.7+) already does captive portal checks. And I observed some detection failure when `network.trr.mode = 3` and `network.trr.wait-for-portal = true`. This happens very occasionally, so I am not sure.

I confirm the patched code now works on both Firefox 60 and Firefox Dev Preview 61. A working configuration is listed here:
```
// I hate 20-minute blacklist
network.trr.blacklist-duration;5

// My DoH servers have multiple addresses so I leave empty and rely on /etc/hosts
network.trr.bootstrapAddress;

// Disable DoH detection
network.trr.confirmationNS;skip

// I am on an IPv6 environment and I prefer IPv6
network.trr.early-AAAA;true

// I didn't set bootAddress so I have to make this 2
// If it is 2, DNS queries will be plaintext during the first seconds you start Firefox
// But I deployed DoH system-wide, so I don't care
network.trr.mode;2

// Either true or false is okay.
// If you set true, you will have better cache if you use HTTP cache,
// but your requests will be logged by the server if verbose=true
network.trr.useGET;true

// I disable this for the reason I described
network.trr.wait-for-portal;false
```

Just to mention that, I recommend you to use a use a server load balancer (e.g. Caddy, Nginx) and set up TLS there, because this program does not do OCSP Stapling, which is necessary for client bootstrapping in a network environment with completely no traditional DNS service.
